// systeminformation/lib/internet.js
'use strict';
// ==================================================================================
// internet.js
// ----------------------------------------------------------------------------------
// Description:   System Information - library
//                for Node.js
// Copyright:     (c) 2014 - 2018
// Author:        Sebastian Hildebrandt
// ----------------------------------------------------------------------------------
// License:       MIT
// ==================================================================================
// 12. Internet
// ----------------------------------------------------------------------------------

var exec = require('child_process').exec;
var util = require('./util');

var _platform = process.platform;

var _linux = (_platform === 'linux');
var _darwin = (_platform === 'darwin');
var _windows = (_platform === 'win32');
var _freebsd = (_platform === 'freebsd');
var _openbsd = (_platform === 'openbsd');
var _sunos = (_platform === 'sunos');

var opts = {
  windowsHide: true
};

// --------------------------
// check if external site is available

function inetChecksite(url, callback) {

  return new Promise(function(resolve) {
    process.nextTick(function() {

      var result = {
        url: url,
        ok: false,
        status: 404,
        ms: -1
      };
      if (url) {
        url = url.toLowerCase();
        var t = Date.now();
        if (_linux || _freebsd || _openbsd || _darwin || _sunos) {
          var args = ' -I --connect-timeout 5 -m 5 ' + url + ' 2>/dev/null | head -n 1 | cut -d " " -f2';
          var cmd = 'curl';
          exec(cmd + args, function (error, stdout) {
            var statusCode = parseInt(stdout.toString());
            result.status = statusCode || 404;
            result.ok = !error && (statusCode === 200 || statusCode === 301 || statusCode === 302 || statusCode === 304);
            result.ms = (result.ok ? Date.now() - t : -1);
            if (callback) { callback(result); }
            resolve(result);
          });
        }
        if (_windows) {   // if this is stable, this can be used for all OS types
          var http = (url.startsWith('https:') ? require('https') : require('http'));
          try {
            http.get(url, function(res) {
              var statusCode = res.statusCode;

              result.status = statusCode || 404;
              result.ok = (statusCode === 200 || statusCode === 301 || statusCode === 302 || statusCode === 304);

              if (statusCode !== 200) {
                res.resume();
                result.ms = (result.ok ? Date.now() - t : -1);
                if (callback) { callback(result); }
                resolve(result);
              } else {
                // res.on('data', function(chunk) {  });
                res.on('end', function() {
                  result.ms = (result.ok ? Date.now() - t : -1);
                  if (callback) { callback(result); }
                  resolve(result);
                });
              }
            }).on('error', function() {
              if (callback) { callback(result); }
              resolve(result);
            });
          } catch (err) {
            if (callback) { callback(result); }
            resolve(result);
          }
        }
      } else {
        if (callback) { callback(result); }
        resolve(result);
      }
    });
  });
}

exports.inetChecksite = inetChecksite;

// --------------------------
// check inet latency

function inetLatency(host, callback) {

  // fallback - if only callback is given
  if (util.isFunction(host) && !callback) {
    callback = host;
    host = '';
  }

  host = host || '8.8.8.8';

  return new Promise(function(resolve) {
    process.nextTick(function() {
      var cmd;
      if (_linux || _freebsd || _openbsd || _darwin) {
        if (_linux) {
          cmd = 'ping -c 2 -w 3 ' + host + ' | grep rtt';
        }
        if (_freebsd || _openbsd) {
          cmd = 'ping -c 2 -t 3 ' + host + ' | grep round-trip';
        }
        if (_darwin) {
          cmd = 'ping -c 2 -t 3 ' + host + ' | grep avg';
        }

        exec(cmd, function (error, stdout) {
          var result = -1;
          if (!error) {
            var line = stdout.toString().split('=');
            if (line.length > 1) {
              var parts = line[1].split('/');
              if (parts.length > 1) {
                result = parseFloat(parts[1]);
              }
            }
          }
          if (callback) { callback(result); }
          resolve(result);
        });
      }
      if (_sunos) {
        exec('ping -s -a ' + host + ' 56 2 | grep avg', {timeout: 3000}, function (error, stdout) {
          var result = -1;
          if (!error) {
            var line = stdout.toString().split('=');
            if (line.length > 1) {
              var parts = line[1].split('/');
              if (parts.length > 1) {
                result = parseFloat(parts[1].replace(',', '.'));
              }
            }
          }
          if (callback) { callback(result); }
          resolve(result);
        });
      }
      if (_windows) {
        var result = -1;
        try {
          exec('ping ' + host + ' -n 1', opts, function (error, stdout) {
            if (!error) {
              var lines = stdout.toString().split('\r\n');
              lines.shift();
              lines.forEach(function (line) {
                if (line.toLowerCase().startsWith('    min')) {
                  var l = line.replace(/ +/g, ' ').split(' ');
                  if (l.length > 8) {
                    result = parseFloat(l[9]);
                  }
                }
              });
            }
            if (callback) { callback(result); }
            resolve(result);
          });  
        } catch (e) {
          if (callback) { callback(result); }
          resolve(result);
        }
      }
    });
  });
}

exports.inetLatency = inetLatency;
