// cocos-utils/cfg/consts.js
module.exports = {
    F_VERSION : "version",
    F_HELP : "help",
    F_NEW : "new",
    F_BUILD : "build",
    F_PUBLISH : "publish",
    F_GEN_RES : "genRes",
    F_GEN_JS_RES : "genJsRes",
    F_GEN_BASE_CFG : "genBaseCfg",
    F_INSTALL : "install",
    F_UPDATE : "update",

    OPT_P : "-p",
    OPT_EM : "-m",
    OPT_T : "-t",
    OPT_FULL : "-full",
    OPT_O : "-o",
    OPT_DC : "-dc",
    OPT_REGISTRY : "-r",
    OPT_VERSION : "-v",

    TEMP_NAME : "tempName",
    DIR : "dir",
    FULL : "full",



    DEFAULT_PROJ_TEMP : "project",
    DEFAULT_ENGINE_MODULES_PATH : "../node_modules/",

    TEMP_BUILD_XML_PATH : "templates/compile/build.xml",

    TEMP_INSTALL_PACKAGE_PATH : "templates/install/package.json",

    RES_4_G_MODULES : "cc.res4GameModules",
    MAIN_PATH : "main.js",
    RES_JS_PATH : "cfg/res.js",
    JS_RES_JS_PATH : "cfg/jsRes.js",
    RES_CFG_JS_PATH : "cfg/resCfg.js",
    BASE_CFG_PATH : "baseCfg.js",
    COCOS_JSON : "cocos.json",
    PKG_JSON : "package.json",
    TPM_PATH : "node_modules",//path of third party modules
    EM_PATH : "../node_modules",//path of engine modules, default "../node_modules"
    TEMP_PATH : "__temp",
    CFG_MERGED_JS : "cfgMerged.js",
    BASE_CFG : "baseCfg",
    BASE_JS_LIST : "baseJsList",
    MODULES_PATH_MAP : "modluesPathMap"
}
